arrows/vxl/convert_image.cxx

```diff
@@ -0,0 +1,508 @@
+                            vil_image_view< bool >& dst,
+                            unsigned sampling_points )
+{
+  // Calculate thresholds
+  std::vector< PixType > thresholds;
+  get_image_percentiles( src, percentiles, thresholds, sampling_points );
+  dst.set_size( src.ni(), src.nj(), percentiles.size() );
+
+  // Perform thresholding
+  for( unsigned i = 0; i < thresholds.size(); i++ )
+  {
+    vil_image_view< bool > output_plane = vil_plane( dst, i );
+    vil_threshold_above( src, output_plane, thresholds[ i ] );
+  }
+}
+
+template < typename InputType, typename OutputType >
+void
+percentile_scale_image( const vil_image_view< InputType >& src,
+                        vil_image_view< OutputType >& dst,
+                        double lower, double upper,
+                        unsigned sampling_points,
+                        bool ignore_extremes = true )
+{
+  std::vector< double > percentiles( 2, 0.0 );
+  percentiles[ 0 ] = lower;
+  percentiles[ 1 ] = upper;
+
+  std::vector< InputType > per_values;
+
+  get_image_percentiles( src, percentiles, per_values, sampling_points,
+                         ignore_extremes );
+
+  OutputType max_val = std::numeric_limits< OutputType >::max();
+
+  InputType lower_bound = per_values[ 0 ];
+  InputType upper_bound = per_values[ 1 ];
+
+  double scale;
+
+  if( per_values[ 1 ] - per_values[ 0 ] > 0 )
+  {
+    scale = ( static_cast< double >( max_val ) + 0.5 ) /
+            ( per_values[ 1 ] - per_values[ 0 ] );
+  }
+  else
+  {
+    scale = static_cast< double >( max_val ) /
+            std::numeric_limits< InputType >::max();
+  }
+
+  unsigned ni = src.ni(), nj = src.nj(), np = src.nplanes();
+  dst.set_size( ni, nj, np );
+
+  std::ptrdiff_t istep = src.istep(), jstep = src.jstep(),
+                 pstep = src.planestep();
+
+  const InputType* plane = src.top_left_ptr();
+  for( unsigned p = 0; p < np; ++p, plane += pstep )
+  {
+    const InputType* row = plane;
+    for( unsigned j = 0; j < nj; ++j, row += jstep )
+    {
+      const InputType* pixel = row;
+      for( unsigned i = 0; i < ni; ++i, pixel += istep )
+      {
+        if( *pixel < lower_bound )
+        {
+          dst( i, j, p ) = 0;
+        }
+        else if( *pixel > upper_bound )
+        {
+          dst( i, j, p ) = std::numeric_limits< OutputType >::max();
+        }
+        else
+        {
+          dst( i, j,
+               p ) =
+            static_cast< OutputType >( ( *pixel - lower_bound ) *
+                                       scale );
+        }
+      }
+    }
+  }
+}
+
+} // end anonoymous namespace
+
+// ----------------------------------------------------------------------------
+/// Private implementation class
+class convert_image::priv
+{
+public:
+
+  priv()
+    : format( "byte" )
+      , single_channel( false )
+      , scale_factor( 0.0 )
+      , random_greyscale( 0.0 )
+      , percentile_norm( -1.0 )
+  {
+  }
+
+  ~priv()
```

Comment:
pedantic: an explicitly empty dtor in a class defined in a `.cxx` is rather pointless